temporal-sdk/src/test/java/io/temporal/workflow/updateTest/UpdateWithStartTest.java

```diff
@@ -0,0 +1,464 @@
+ *
+ * Copyright (C) 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this material except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.temporal.workflow.updateTest;
+
+import static org.junit.Assert.*;
+
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
+import io.temporal.api.common.v1.WorkflowExecution;
+import io.temporal.api.enums.v1.WorkflowIdConflictPolicy;
+import io.temporal.api.enums.v1.WorkflowIdReusePolicy;
+import io.temporal.client.*;
+import io.temporal.testing.internal.SDKTestOptions;
+import io.temporal.testing.internal.SDKTestWorkflowRule;
+import io.temporal.worker.WorkerOptions;
+import io.temporal.workflow.*;
+import io.temporal.workflow.shared.TestMultiArgWorkflowFunctions;
+import io.temporal.workflow.shared.TestWorkflows;
+import java.util.*;
+import java.util.concurrent.ExecutionException;
+import java.util.function.BiFunction;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class UpdateWithStartTest {
+
+  @Rule
+  public SDKTestWorkflowRule testWorkflowRule =
+      SDKTestWorkflowRule.newBuilder()
+          .setWorkerOptions(WorkerOptions.newBuilder().build())
+          .setWorkflowTypes(
+              WorkflowWithUpdateImpl.class,
+              TestUpdatedWorkflowImpl.class,
+              TestMultiArgWorkflowFunctions.TestMultiArgWorkflowImpl.class)
+          .build();
+
+  @Test
+  public void startAndSendUpdateTogether() throws ExecutionException, InterruptedException {
+    WorkflowClient workflowClient = testWorkflowRule.getWorkflowClient();
+
+    WorkflowOptions options = createOptions();
+    TestWorkflows.WorkflowWithUpdate workflow =
+        workflowClient.newWorkflowStub(TestWorkflows.WorkflowWithUpdate.class, options);
+
+    WorkflowStartOperationUpdate<String> updateOp =
+        WorkflowStartOperationUpdate.newBuilder(workflow::update, 1, "Hello Update")
+            .setWaitForStage(WorkflowUpdateStage.COMPLETED)
+            .build();
+
+    WorkflowUpdateHandle<String> handle1 =
+        WorkflowClient.startWithOperation(workflow::execute, updateOp);
+    assertEquals(options.getWorkflowId(), handle1.getExecution().getWorkflowId());
+    assertEquals("Hello Update", handle1.getResultAsync().get());
+
+    WorkflowUpdateHandle<String> updHandle = updateOp.getUpdateHandle().get();
+    assertEquals(updateOp.getResult(), updHandle.getResultAsync().get());
```

Comment:
I wonder if the `WorkflowUpdateHandle` should also have `getResult`()?

Reply:
> I wonder if the WorkflowUpdateHandle should also have getResult()?

Would there be a meaningful difference between `getResult()` and `getResultAsync()` besides that the former just waits on the future of the latter? Also, @Quinn-With-Two-Ns are we sure we want to suffix the word `Async` to `getResult`? And if we do, should `WorkflowStartOperationUpdate.getUpdateHandle` be renamed to `WorkflowStartOperationUpdate.getUpdateHandleAsync` for consistency?

Reply:
Is there still a use case for getting a handle from this operation if `updateWithStart` is changed to return an update handle?

Reply:
Yes, I think getting a handle always has value so you can pass it around, so you can see the update ID, and so that you can do other things on it once we add them (I hope we add "describe" soon, CLI already saw a need for it).

Reply:
Correct, there is probably not value. It doesn't make too much sense to have two ways to get a handle.

Reply:
Only reason I could think of is as a promise, e.g. to give you something to wait on before the start call even starts, e.g.

```java
public Optional<WorkflowStartOperationUpdate<String>> addUpdateIfNeeded(SomeParam someParam) {
  if (!someParam.isUpdateNeeded()) {
    return Optional.empty();
  }
  var update = WorkflowStartOperationUpdate.newBuilder(workflow::update, 1, "Hello Update")
        .setWaitForStage(WorkflowUpdateStage.COMPLETED)
        .build();
  this.asynchronouslyDealWithUpdateResult(update.getUpdateHandle());
  return Optional.of(update);
}
```

I could derive a few helpers, but suffice to say it is a use case to be able to await on a future once it's created, not once it's applied. Having said that we could probably get away with not having it.